dll/win32/shdocvw/CNSCBand.cpp

```diff
@@ -0,0 +1,1682 @@
+        if (tmp)
+        {
+            hItem = tmp;
+            continue;
+        }
+
+        if (bInsert)
+        {
+            *phItem = hItem = _InsertItem(hParent, dest, ILFindLastID(dest), TRUE);
+            return TRUE;
+        }
+
+        return FALSE;
+    }
+
+    UNREACHABLE;
+}
+
+BOOL CNSCBand::_NavigateToCurrentFolder()
+{
+    CComHeapPtr<ITEMIDLIST> pidl;
+    HRESULT hr = _GetCurrentLocation(&pidl);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return FALSE;
+
+    // Find PIDL into our explorer
+    ++m_mtxBlockNavigate;
+    HTREEITEM hItem;
+    BOOL result = _NavigateToPIDL(pidl, &hItem, TRUE, FALSE, TRUE);
+    --m_mtxBlockNavigate;
+
+    return result;
+}
+
+// *** message handlers ***
+
+LRESULT CNSCBand::OnCreate(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    INITCOMMONCONTROLSEX iccx = { sizeof(iccx), ICC_TREEVIEW_CLASSES | ICC_BAR_CLASSES };
+    if (!::InitCommonControlsEx(&iccx))
+        return -1;
+    if (FAILED_UNEXPECTEDLY(_CreateToolbar()) || FAILED_UNEXPECTEDLY(_CreateTreeView()))
+        return -1;
+    return 0;
+}
+
+LRESULT CNSCBand::OnDestroy(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    m_hwndTreeView.DestroyWindow();
+    m_hwndToolbar.DestroyWindow();
+    return 0;
+}
+
+LRESULT CNSCBand::OnSize(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL& bHandled)
+{
+    if (!m_hwndTreeView)
+        return 0;
+
+    RECT rc;
+    GetClientRect(&rc);
+    LONG cx = rc.right, cy = rc.bottom;
+
+    RECT rcTB;
+    LONG cyTB = 0;
+    if (m_hwndToolbar)
+    {
+        m_hwndToolbar.SendMessage(TB_AUTOSIZE, 0, 0);
+        m_hwndToolbar.GetWindowRect(&rcTB);
+        cyTB = rcTB.bottom - rcTB.top;
+    }
+
+    m_hwndTreeView.MoveWindow(0, cyTB, cx, cy - cyTB);
+    return 0;
+}
+
+LRESULT CNSCBand::OnSetFocus(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    m_bFocused = TRUE;
+    IUnknown_OnFocusChangeIS(m_pSite, reinterpret_cast<IUnknown*>(this), TRUE);
+    bHandled = FALSE;
+    return 0;
+}
+
+LRESULT CNSCBand::OnKillFocus(UINT uMsg, WPARAM wParam, LPARAM lParam, BOOL &bHandled)
+{
+    IUnknown_OnFocusChangeIS(m_pSite, reinterpret_cast<IUnknown*>(this), FALSE);
+    m_bFocused = FALSE;
+    return 0;
+}
+
+HRESULT CNSCBand::_AddFavorite()
+{
+    CComHeapPtr<ITEMIDLIST> pidlCurrent;
+    _GetCurrentLocation(&pidlCurrent);
+
+    WCHAR szCurDir[MAX_PATH];
+    if (!ILGetDisplayName(pidlCurrent, szCurDir))
+    {
+        FIXME("\n");
+        return E_FAIL;
+    }
+
+    WCHAR szPath[MAX_PATH], szSuffix[32];
+    SHGetSpecialFolderPathW(m_hWnd, szPath, CSIDL_FAVORITES, TRUE);
+    PathAppendW(szPath, PathFindFileNameW(szCurDir));
+
+    const INT ich = lstrlenW(szPath);
+    for (INT iTry = 2; iTry <= 9999; ++iTry)
+    {
+        PathAddExtensionW(szPath, L".lnk");
+        if (!PathFileExistsW(szPath))
+            break;
+        szPath[ich] = UNICODE_NULL;
+        wsprintfW(szSuffix, L" (%d)", iTry);
+        lstrcatW(szPath, szSuffix);
+    }
```

Comment:
What is that strange loop with the ~10000 retries?

Reply:
This is a renaming loop.